packages/loader/container-loader/src/container.ts

```diff
@@ -2045,7 +2050,7 @@ export class Container
 
 	private processSignal(message: ISignalMessage) {
 		// No clientId indicates a system signal message.
-		if (message.clientId === null) {
+		if (protocolHandlerShouldProcessSignal(message)) {
```

Comment:
ContainerContext handles client created signals so there can be many signal types there. ProtocolHandler is the one that has well defined and a limited number of signals to handle so this is the handler we should be string on.

Reply:
I think ideally we should prescribe exactly the types of messages that should go to both and blow up on anything unrecognized.  And also ideally this is done with an explicit `type` rather than hints like null clientId.  We're probably somewhat limited due to backcompat here though.

Reply:
For what it's worth: we are moving to a world where ALL ops are delivered to runtime.
Reason for this move - I want to (over time) reduce exposure of information in IDeltaManager and runtime reaching out to loader. And allow runtime to present a "lie" to layers below if it needs to (like op virtualization).
Making it happened requires a lot of changes (and slow deployment).

I do not know if we want to do same for signals - there does not seem like as good story here. But maybe we want to be symmetrical.

Reply:
I can't think of a reason why we would explicitly _not_ want signals to go to the runtime; symmetry does feel like the more appealing route. Not sure if the ClientJoin/ClientLeave signals (and Clear, which seems like a somewhat special case?) would be a good fit for the runtime; if they are, we can get to full symmetry, and if not, we can keep handling those as exceptions in the loader and let everything else flow to the runtime.

> I think ideally we should prescribe exactly the types of messages that should go to both and blow up on anything unrecognized.

Isn't that a very hard stop for some back-compat scenarios? Introducing a new signal type would absolutely require no old loaders around anymore, right? VS just letting the loader handle anything it explicitly knows/wants to and let everything else flow to the runtime. Unless the explicit `type` you're proposing is just "loader vs runtime", and any specifics of new signals go in some other property or the payload; I can see that world (although then we could ask "do we foresee having to route signals to some third place at some point, and this becomes a back-compat blocker to _that_ (if the loader blows up on unknown types)").

@ghoshkaj I think Vlad's comment is a point in favor of changing

Reply:
> Isn't that a very hard stop for some back-compat scenarios? Introducing a new signal type would absolutely require no old loaders around anymore, right? VS just letting the loader handle anything it explicitly knows/wants to and let everything else flow to the runtime. Unless the explicit type you're proposing is just "loader vs runtime", and any specifics of new signals go in some other property or the payload; I can see that world (although then we could ask "do we foresee having to route signals to some third place at some point, and this becomes a back-compat blocker to that (if the loader blows up on unknown types)").

Yes, I want the "type" to just be routing info - more or less the envelope concept.  And so also yes - if a third thing is required and the router doesn't know how to find it, then I want it to blow up.  The loader concept needs to be split differently (e.g. part of it is what Vlad mentions about op handling), some of this backcompat threat probably becomes easier then.

> I do not know if we want to do same for signals - there does not seem like as good story here. But maybe we want to be symmetrical.

My gut is that the audience shouldn't live in the runtime.  Its messages and concepts are service-specific (e.g. the IUser being typed to the service), I'd rather see it as a peer object to the runtime that we route the messages to.  With the goal being that the runtime can remain "cross-platform" by default.  The customer can still inject it into the runtime if they choose, but I don't think it needs to be there by default.

For the same reason I'm not sure about "all" ops being delivered to the runtime.  I like the goals you state but I'm not sure all ops are best handled by the runtime vs. other objects.  For example, the Quorum (and its ops), code proposal, and code loading might be better split off into some bootloader object (separate from both the loader and the runtime).  Probably worth some deeper design discussion.